frontends/openssl-provider/common.c

```diff
@@ -0,0 +1,406 @@
+/* RAND parameter descriptors */
+static OSSL_FUNC_rand_gettable_ctx_params_fn esdm_rand_gettable_ctx_params;
+/* RAND parameters */
+static OSSL_FUNC_rand_get_ctx_params_fn esdm_rand_get_ctx_params;
+
+static void *esdm_rand_newctx(void *provctx, void *parent __unused,
+                              const OSSL_DISPATCH *parent_calls __unused)
+{
+  struct esdm_provider_ctx *cprov = provctx;
+  struct esdm_rand_ctx *rand = OPENSSL_secure_zalloc(sizeof(struct esdm_rand_ctx));
+
+  if (rand == NULL) {
+    goto err;
+  }
+
+  rand->core = cprov->core;
+  return rand;
+
+err:
+  OPENSSL_secure_clear_free(rand, sizeof(struct esdm_rand_ctx));
+  return NULL;
+}
+
+static void esdm_rand_freectx(void *ctx)
+{
+  struct esdm_rand_ctx *rand = ctx;
+
+  if (rand == NULL)
+    return;
+
+  CRYPTO_THREAD_lock_free(rand->lock);
+  OPENSSL_secure_clear_free(rand, sizeof(struct esdm_rand_ctx));
+}
+
+static int esdm_rand_instantiate(void *ctx __unused,
+                                 unsigned int strength __unused,
+                                 int prediction_resistance __unused,
+                                 const unsigned char *pstr __unused,
+                                 size_t pstr_len __unused,
+                                 const OSSL_PARAM params[] __unused)
+{
+  return 1;
+}
+
+static int esdm_rand_uninstantiate(void *ctx __unused)
+{
+  return 1;
+}
+
+static int esdm_rand_generate(void *ctx __unused, unsigned char *out,
+                              size_t outlen, unsigned int strength __unused,
+                              int prediction_resistance,
+                              const unsigned char *addin, size_t addin_len)
+{
+  ssize_t ret;
+
+  if (!out)
+    goto err;
+
+  if (addin && addin_len > 0) {
+    esdm_invoke(esdm_rpcc_write_data(addin, addin_len));
+    if (ret != 0)
+      goto err;
+  }
+
+  if (prediction_resistance) {
+    esdm_invoke(esdm_rpcc_get_random_bytes_pr(out, outlen));
+  } else {
+    esdm_invoke(esdm_rpcc_get_random_bytes_full(out, outlen));
+  }
+  if (ret != (ssize_t)outlen)
+    goto err;
+
+  return 1;
+
+err:
+  return 0;
+}
+
+static int esdm_rand_reseed(void *ctx __unused,
+                            int prediction_resistance __unused,
+                            const unsigned char *ent, size_t ent_len,
+                            const unsigned char *addin, size_t addin_len)
+{
+  ssize_t ret;
+
+  /* unaccounted writing of additional data does no harm */
+  if (ent && ent_len > 0)
+    esdm_invoke(esdm_rpcc_write_data(ent, ent_len));
+  if (addin && addin_len > 0)
+    esdm_invoke(esdm_rpcc_write_data(addin, addin_len));
+
+  return 1;
+}
+
+static size_t esdm_rand_nonce(void *ctx __unused, unsigned char *out,
+                              unsigned int outlen, size_t min_noncelen __unused,
+                              size_t max_noncelen __unused)
+{
+  ssize_t ret;
+
+  if (out == NULL)
+    return outlen;
+
+  esdm_invoke(esdm_rpcc_get_random_bytes_full(out, outlen));
```

Comment:
Perhaps get_random_bytes_min would suffice?

Reply:
I think get_random_bytes_min should also be sufficient here, as this is only used as a nonce additional to entropy, when instantiating DRBGs in OpenSSL.